Author: 
Description: 
Index: coreutils/src/id.c
===================================================================
--- coreutils.orig/src/id.c
+++ coreutils/src/id.c
@@ -380,6 +380,10 @@ print_full_info (char const *username)
         ok &= false;
         return;
       }
+    else if (sysconf(_SC_NGROUPS_MAX) > 0 && n_groups > sysconf(_SC_NGROUPS_MAX))
+      {
+        fprintf (stderr, _("Warning: user %s is in more groups than system's configured maximum.\n"), (username != NULL)?username:"");
+      }
 
     if (n_groups > 0)
       fputs (_(" groups="), stdout);
